### ISSUE TITLE
[Classic] De-aggro no longer working properly

### ISSUE THREAD

**Issue by nekolepsy** [cmangos/issues#1585]
**Current behavior**: 
Getting out of combat by running away is almost impossible. Enemies will just keep following you. I've managed to eventually out run things on a mount or sometimes when crossing rivers.

**Expected behavior**: 
Be able to get out of combat after running for a short period of time.

**Steps to reproduce**:

1. Aggro something
2. Run away from it 

**Client version**:  1.12.1

**Commit hash**: df94350d9333fe129eea2f79882841cfe6b9f660

**Database version**: classic-db 6053d376152353ca10d04070ecf5f289d32c268a

**Operating system**: Centos 7

I decided to upgrade my test server today and found this issue.

**Comment by cala** [comment]
Confirmed: leashing is broken since the last batch of commits... Was this even tested?

**Comment by killerwife** [comment]
I will check out what changed in the leashing code. I wasn't aware that I deleted some distance checks.

**Comment by SilvioDoMine** [comment]
As soon as you get into it, @killerwife, could you message me with a fix?

**Comment by killerwife** [comment]
@cala I am gonna need your help here. How did it work before my batch of commits. The reason why I am asking is that on voa we have a prototype, which I can if need be, merge to cmangos. The real impl is waiting for something from WB. Though if I find a reason what changed, I can fix it until then easily.

**Comment by cala** [comment]
Unfortunately, I can't tell you how the mechanics was implemented in the core, if this is what your are asking for.

I can describe you a few examples of how it was working in-game though and provide a few testing examples.

**Comment by killerwife** [comment]
I just want to know if the mob evaded before my this weeks commit. Because I have gone over every commit of mine, and it shouldn't affect any evading due to distance/chasing whatsoever.

**Comment by cala** [comment]
Reverting to https://github.com/cmangos/mangos-classic/commit/ee327422ee5aa234b020b5e405322e8211af87f9 (which is the last commit before your batch) re-introduce a functional leashing system.

> Because I have gone over every commit of mine, and it shouldn't affect any evading due to distance/chasing whatsoever.

Maybe it is the same magical commit than the one that fixed the issue with trap GOs with no faction ? 😉

**Comment by killerwife** [comment]
Thx for the assist. I guess its time for actual reverting on my end. I will do it today/tomorrow.

**Comment by cala** [comment]
Here is your culprit : https://github.com/cmangos/mangos-classic/commit/b5ba0942f2e64f9e0329f84f245e209557ecc897 and is **not** fixed by https://github.com/cmangos/mangos-classic/commit/4ccb71680990ea264a8cb21f563abe1629b4292c
I found it with 5 iterations of dichotomy but were I to have a guess, it would have been my first try nonetheless.

From what I understand after quickly looking at the commit, you want to modify the way all the creatures/units (thousands of them) are handling threat and selecting their target to match the behaviour of only 4 creatures (in fact, this may be at least 2 more from Classic, but that's irrelevant here).

Those few creatures are all handled in SD2 with a working code for this behaviour (at least Ragnaros, but the logic could be the same with the others), so I don't see the need of changing all creatures. If your purpose is to remove the specific threat code from SD2 in order to have the script look straight like some kind of ACID script, there may be another way to move it rather than making a generic rule from an exception?

**Comment by killerwife** [comment]
The function you see there is essentially AI behaviour, which should also be in AI (like selecthostiletarget), but yet isn't. The reason I added it here, is because the way it was handled in SD2 was not sufficient without 50+ copy pasted lines. Also another fact, before I started these commits, creature was not capable of nullifying its target inside combat, for the purpose of casting at correct targets and mimicking human behaviour.

Don't get me wrong, all of what I am doing (look at the channeling commit thats close to this one) is case by case analysis and extension. The problem is that the amount of mobs I am encountering with this is increasing, and yes in vanilla its only one mob, but there seem to be more in TBC, and even more in wotlk. The ideal case is to move the whole decision making out of threat manager, to AI.

The reason why it needs to be in AI, is that for example Ranged Spell school immunity on target, is a factor in deciding target. And there are plenty more, this "ignore ranged targets" is only one of them, and the first attempt at what in the future will most likely be threat prioritization based on factors like stun, fear, confuse, immunity, and so on.

That being said, I will fix this issue very soon.
